client/graphics/vertex-buffer-object.cpp

```diff
@@ -1,173 +1,175 @@
+			for (ObjectGroup* objgroup : mfmesh->object_groups) {
+				for (auto mat_group : objgroup->material_groups) {
+					VertexGroup group;
+					group.index_count = mat_group.count;
+					group.starting_offset = mat_group.start + all_indices.size();
+					group.mesh_group_number = i;
+					if (MaterialMap::Has(mat_group.material_name)) {
+						group.material = MaterialMap::Get(mat_group.material_name);
+					}
+					else {
+						group.material = Material::Create(mat_group.material_name);
+						group.material->SetDrawElementsMode(GL_TRIANGLES);
+						group.material->SetPolygonMode(GL_FILL);
+						for (auto texture : mat_group.textures) {
+							if (TextureMap::Has(texture)) {
+								group.material->AddTexture(TextureMap::Get(texture));
 							}
 						}
-						this->vertex_groups.push_back(std::move(group));
-					}
-					for (GLuint index : objgroup->indices) {
-						all_indices.push_back(index + vert_offset);
 					}
+					this->vertex_groups.push_back(std::move(group));
+				}
+				for (GLuint index : objgroup->indices) {
+					all_indices.push_back(index + vert_offset);
 				}
 			}
-			Load(all_verts, all_indices);
 		}
+		Load(all_verts, all_indices);
 	}
+}
 
-	void VertexBufferObject::Load(const std::vector<VertexData>& verts, const std::vector<GLuint>& indices) {
-		if (!this->vao) {
-			glGenVertexArrays(1, &this->vao);
-		}
-		if (!this->vbo) {
-			glGenBuffers(1, &this->vbo);
-		}
-		if (!this->ibo) {
-			glGenBuffers(1, &this->ibo);
-		}
+void VertexBufferObject::Load(const std::vector<VertexData>& verts, const std::vector<GLuint>& indices) {
+	if (!this->vao) {
+		glGenVertexArrays(1, &this->vao);
+	}
+	if (!this->vbo) {
+		glGenBuffers(1, &this->vbo);
+	}
+	if (!this->ibo) {
+		glGenBuffers(1, &this->ibo);
+	}
 
-		glBindVertexArray(this->vao);
-		glBindBuffer(GL_ARRAY_BUFFER, this->vbo);
-		// If the size hasn't changed we can call update
-		if (this->vertex_count >= verts.size()) {
-			auto* buffer = glMapBufferRange(GL_ARRAY_BUFFER, 0, this->vertex_count *
-											sizeof(VertexData), GL_MAP_WRITE_BIT | GL_MAP_UNSYNCHRONIZED_BIT | GL_MAP_INVALIDATE_BUFFER_BIT);
-			if (buffer) {
-				std::memcpy(buffer, &verts[0], verts.size() * sizeof(VertexData));
-				std::memset((char*)buffer + verts.size() * sizeof(VertexData) - 1, 0, (this->vertex_count - verts.size()) * sizeof(VertexData));
-				glUnmapBuffer(GL_ARRAY_BUFFER);
-			}
-			else {
-				//std::err << "glMapBufferRange() failed" << __LINE__ << __FILE__ << std::endl;
-			}
+	glBindVertexArray(this->vao);
+	glBindBuffer(GL_ARRAY_BUFFER, this->vbo);
+	// If the size hasn't changed we can call update
+	if (this->vertex_count >= verts.size()) {
+		auto* buffer = glMapBufferRange(
+				GL_ARRAY_BUFFER,
+				0,
+				this->vertex_count * sizeof(VertexData),
+				GL_MAP_WRITE_BIT | GL_MAP_UNSYNCHRONIZED_BIT | GL_MAP_INVALIDATE_BUFFER_BIT);
+		if (buffer) {
+			std::memcpy(buffer, &verts[0], verts.size() * sizeof(VertexData));
```

Comment:
verts.size() * sizeof(VertexData) is used a few times and could be stored in a variable.